In [1]:
from sqlalchemy import create_engine
import pandas as pd

In [2]:
database_name = 'scooters'    # Fill this in with your scooter database name

connection_string = f"postgresql://postgres:postgres@localhost:5432/{database_name}"

Set up is complete, next is the engine

In [3]:
engine = create_engine(connection_string)

The next part is the beginning of the SQL queries

In [4]:
query = '''
SELECT *
FROM INFORMATION_SCHEMA.COLUMNS
WHERE TABLE_NAME = 'scooters'
ORDER BY ORDINAL_POSITION
'''

result = engine.execute(query)
lat = pd.read_sql(query, con = engine)
lat.head()

table_catalog table_schema table_name  column_name  ordinal_position   
0      scooters       public   scooters  pubdatetime                 1  \
1      scooters       public   scooters     latitude                 2   
2      scooters       public   scooters    longitude                 3   
3      scooters       public   scooters       sumdid                 4   
4      scooters       public   scooters     sumdtype                 5   

  column_default is_nullable                    data_type   
0           None         YES  timestamp without time zone  \
1           None         YES                      numeric   
2           None         YES                      numeric   
3           None         YES                         text   
4           None         YES                         text   

  character_maximum_length  character_octet_length  ... is_identity   
0                     None                     NaN  ...          NO  \
1                     None                     NaN  ...          NO   
2                     None                     NaN  ...          NO   
3                     None            1.073742e+09  ...          NO   
4                     None            1.073742e+09  ...          NO   

   identity_generation identity_start  identity_increment identity_maximum   
0                 None           None                None             None  \
1                 None           None                None             None   
2                 None           None                None             None   
3                 None           None                None             None   
4                 None           None                None             None   

  identity_minimum identity_cycle is_generated generation_expression   
0             None             NO        NEVER                  None  \
1             None             NO        NEVER                  None   
2             None             NO        NEVER                  None   
3             None             NO        NEVER                  None   
4             None             NO        NEVER                  None   

  is_updatable  
0          YES  
1          YES  
2          YES  
3          YES  
4          YES  

[5 rows x 44 columns]

In [5]:
query = '''
SELECT *
FROM scooters
LIMIT 5;
'''

result = engine.execute(query)
lat = pd.read_sql(query, con = engine)
lat.head()

pubdatetime   latitude  longitude         sumdid sumdtype   
0 2019-05-29 06:38:22.663  36.086508 -86.713559  Powered132581  Powered  \
1 2019-05-29 06:38:22.777  36.151483 -86.789033  Powered265244  Powered   
2 2019-05-29 06:38:22.787  36.138829 -86.799535  Powered853835  Powered   
3 2019-05-29 06:38:22.893  36.147757 -86.793564  Powered140629  Powered   
4 2019-05-29 06:38:22.927  36.138345 -86.808485  Powered447401  Powered   

   chargelevel sumdgroup  costpermin companyname  
0         52.0   scooter        0.15        Lyft  
1         44.0   scooter        0.15        Lyft  
2         74.0   scooter        0.15        Lyft  
3         52.0   scooter        0.15        Lyft  
4         80.0   scooter        0.15        Lyft

In [6]:
query2 = '''
SELECT *
FROM trips
LIMIT 20;
'''

result = engine.execute(query2)
lat2 = pd.read_sql(query2, con = engine)
lat2.head()

pubtimestamp companyname triprecordnum         sumdid   
0 2019-06-15 07:34:30.840        Lyft        LFT820  Powered970030  \
1 2019-06-15 07:34:37.330        Lyft        LFT821  Powered646396   
2 2019-06-15 07:35:02.050        Lyft        LFT822  Powered962686   
3 2019-06-15 07:35:58.107        Lyft        LFT823  Powered605278   
4 2019-06-15 07:36:09.570        Lyft        LFT824  Powered284638   

   tripduration  tripdistance   startdate        starttime     enddate   
0      8.405267    8287.40184  2019-06-15  07:26:06.406666  2019-06-15  \
1     32.691783   29301.18204  2019-06-15  07:01:55.720000  2019-06-15   
2     34.421983   29514.43664  2019-06-15  07:00:36.623333  2019-06-15   
3     11.380533   11138.45180  2019-06-15  07:24:35.160000  2019-06-15   
4     51.318883   46725.72328  2019-06-15  06:44:50.333333  2019-06-15   

           endtime  startlatitude  startlongitude  endlatitude  endlongitude   
0  07:34:30.723333       36.17789       -86.79122     36.16502     -86.79226  \
1  07:34:37.226666       36.15306       -86.79613     36.14411     -86.83366   
2  07:35:01.943333       36.15302       -86.79599     36.14415     -86.83418   
3  07:35:57.993333       36.12764       -86.78910     36.13171     -86.77764   
4  07:36:09.466666       36.12762       -86.78907     36.13167     -86.77763   

                                           triproute               create_dt  
0  [(36.17789, -86.79122), (36.17805, -86.79117),... 2019-06-16 08:55:39.063  
1  [(36.15306, -86.79613), (36.153, -86.79622), (... 2019-06-16 08:55:39.097  
2  [(36.15302, -86.79599), (36.153, -86.79594), (... 2019-06-16 08:55:39.127  
3  [(36.12764, -86.7891), (36.12788, -86.78907), ... 2019-06-16 08:55:39.160  
4  [(36.12762, -86.78907), (36.12762, -86.78897),... 2019-06-16 08:55:39.193

- Are there any null values in any columns in either table?
- What date range is represented in each of the date columns? Investigate any values that seem odd.
- Is time represented with am/pm or using 24 hour values in each of the columns that include time?
- What values are there in the sumdgroup column? Are there any that are not of interest for this project?
- What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising?
-What is the range of values for trip duration and trip distance? Do these values make sense? Explore values that might seem questionable.
- Check out how the values for the company name column in the scooters table compare to those of the trips table. What do you notice?

In [7]:
# What is the date range in each of the date columns.   

query4 = '''
SELECT MAX(pubdatetime), MIN(pubdatetime)
FROM scooters;
'''

result = engine.execute(query4)
lat = pd.read_sql(query4, con = engine)
lat.head()

In [ ]:
query5 = '''
SELECT MAX(pubtimestamp), MIN(pubtimestamp)
FROM trips;
'''

result = engine.execute(query5)
lat = pd.read_sql(query5, con = engine)
lat.head()

max                     min
0 2019-08-01 07:04:00 2019-05-01 00:00:55.423

The start and end times do not match between the two tables.  They are off.

In [ ]:
query6 = '''
SELECT MAX(pubdatetime), MIN(pubdatetime)
FROM scooters
WHERE sumdid = 'Powered9EAJL';
'''
result = engine.execute(query6)
lat = pd.read_sql(query6, con = engine)
lat.head()

max                     min
0 2019-05-04 17:05:05 2019-05-01 00:06:41.537

In [ ]:
query7 = '''
SELECT MAX(pubtimestamp), MIN(pubtimestamp)
FROM trips
WHERE sumdid = 'Powered9EAJL';
'''
result = engine.execute(query7)
lat = pd.read_sql(query7, con = engine)
lat.head()

max                     min
0 2019-05-07 18:03:49.303 2019-05-01 00:00:55.423

The first and last trip for Powered9EAJL do not match between tables.

In [ ]:
query8 = '''
SELECT pubdatetime,	latitude, longitude, sumdid, sumdtype, chargelevel, sumdgroup, costpermin, companyname
FROM scooters
WHERE chargelevel IS NULL;
'''

result = engine.execute(query8)
lat = pd.read_sql(query8, con = engine)
lat.head()

pubdatetime   latitude  longitude   
0 2019-07-27 13:51:49  36.167067 -86.781715  \
1 2019-07-27 13:56:49  36.167046 -86.781702   
2 2019-07-27 14:01:49  36.167093 -86.781706   
3 2019-07-27 14:06:49  36.167052 -86.781727   
4 2019-07-28 23:28:05  36.149802 -86.807646   

                                         sumdid sumdtype chargelevel   
0  Powered-e19bf412-f79e-983d-1ebf-410597d7ac5f  Powered        None  \
1  Powered-e19bf412-f79e-983d-1ebf-410597d7ac5f  Powered        None   
2  Powered-e19bf412-f79e-983d-1ebf-410597d7ac5f  Powered        None   
3  Powered-e19bf412-f79e-983d-1ebf-410597d7ac5f  Powered        None   
4  Powered-55a7ac02-7e17-bb3b-130d-b160d9776405  Powered        None   

  sumdgroup  costpermin companyname  
0   Scooter         0.3        Bolt  
1   Scooter         0.3        Bolt  
2   Scooter         0.3        Bolt  
3   Scooter         0.3        Bolt  
4   Scooter         0.3        Bolt

In [ ]:
query9 = '''
SELECT pubtimestamp, companyname, triprecordnum, sumdid, tripduration, tripdistance, 
    startdate, starttime, enddate, endtime, startlatitude, startlongitude, endlatitude, endlongitude, triproute, create_dt
FROM trips
WHERE create_dt IS NULL;
'''

result = engine.execute(query9)
lat = pd.read_sql(query9, con = engine)
lat.head()

Empty DataFrame
Columns: [pubtimestamp, companyname, triprecordnum, sumdid, tripduration, tripdistance, startdate, starttime, enddate, endtime, startlatitude, startlongitude, endlatitude, endlongitude, triproute, create_dt]
Index: []

There are nulls in the chargelevel, but that is about it.


What date range is represented in each of the date columns? Investigate any values that seem odd.

It appears that in all are supposed to be in 24 hour format.


What values are there in the sumdgroup column? Are there any that are not of interest for this project?


In [ ]:
query10 = '''
SELECT DISTINCT sumdgroup
FROM scooters
'''
result = engine.execute(query10)
lat = pd.read_sql(query10, con = engine)
lat.head()

sumdgroup
0   bicycle
1   scooter
2   Scooter

There are also bicycles included in this is for scooters.

What are the minimum and maximum values for all the latitude and longitude columns? Do these ranges make sense, or is there anything surprising?

In [ ]:
query11 = '''
SELECT MAX(longitude)
FROM scooters
'''
result = engine.execute(query11)
lat = pd.read_sql(query11, con = engine)
lat.head()

max
0  0.0

In [ ]:
query13 = '''
SELECT MAX(latitude)
FROM scooters

'''
result = engine.execute(query13)
lat = pd.read_sql(query13, con = engine)
lat.head()

max
0  3.609874e+06

In [ ]:
query14 = '''
SELECT MIN(longitude)
FROM scooters;
'''
result = engine.execute(query14)
lat = pd.read_sql(query14, con = engine)
lat.head()

min
0 -97.443879

In [ ]:
query15 = '''
SELECT MIN(latitude)
FROM scooters;
'''
result = engine.execute(query15)
lat = pd.read_sql(query15, con = engine)
lat.head()

min
0  0.0

In [ ]:
query16 = '''
SELECT latitude, longitude, sumdid, sumdgroup
FROM scooters
WHERE longitude IN (SELECT MIN(longitude)
FROM scooters)
LIMIT 1;

'''
result = engine.execute(query16)
lat = pd.read_sql(query16, con = engine)
lat.head()

latitude  longitude      sumdid sumdgroup
0  35.212984 -97.443879  Powered819   Scooter

In [ ]:
query17 = '''
SELECT latitude, longitude, sumdid, sumdgroup
FROM scooters
WHERE longitude IN (SELECT MAX(longitude)
FROM scooters)
LIMIT 1;
'''
result = engine.execute(query17)
lat = pd.read_sql(query17, con = engine)
lat.head()

latitude  longitude      sumdid sumdgroup
0       0.0        0.0  Powered595   Scooter

It looks like there could be a bike near the equator or at the Prime Meridian, or the GPS is just broken.  And there is another at Oklahoma City Oklahoma.

What is the range of values for trip duration and trip distance? Do these values make sense? Explore values that might seem questionable.

In [ ]:
query18 = '''
SELECT MIN(tripduration)
FROM trips
LIMIT 1
'''
result = engine.execute(query18)
lat = pd.read_sql(query18, con = engine)
lat.head()

min
0 -19.358267

In [ ]:
query19 = '''
SELECT MAX(tripduration)
FROM trips
LIMIT 1
'''
result = engine.execute(query19)
lat = pd.read_sql(query19, con = engine)
lat.head()

max
0  512619.0

There times are not always accurate.  There is one that is negative time, and one that is over 355 days.

What are the min and max distance?

In [ ]:
query20 = '''
SELECT MAX(tripdistance)
FROM trips
LIMIT 1
'''
result = engine.execute(query20)
lat = pd.read_sql(query20, con = engine)
lat.head()

max
0  3.188448e+07

In [ ]:
query21 = '''
SELECT MIN(tripdistance)
FROM trips
LIMIT 1
'''
result = engine.execute(query21)
lat = pd.read_sql(query21, con = engine)
lat.head()

min
0 -20324803.8

The furthest distance is 6,040 miles. The shortest distance is a negative 3,853 miles.

In [ ]:
query22 = '''
SELECT sumdid, tripdistance, tripduration, startdate, starttime, enddate,	endtime, startlatitude, startlongitude, endlatitude, endlongitude, triproute, create_dt
FROM trips
WHERE tripdistance in (
    SELECT MAX(tripdistance)
    FROM trips)
LIMIT 1;

'''
result = engine.execute(query22)
lat = pd.read_sql(query22, con = engine)
lat.head()

sumdid  tripdistance  tripduration   startdate starttime     enddate   
0  Powered2460  3.188448e+07           9.0  2019-07-05  01:18:56  2019-07-05  \

    endtime  startlatitude  startlongitude  endlatitude  endlongitude   
0  01:27:56       36.16338      -86.780142     36.16628    -86.778127  \

                                 triproute               create_dt  
0  [('0', '0'), ('36.16628', '-86.77813')] 2019-07-06 07:13:12.210

In [ ]:
query23 = '''
SELECT sumdid, tripdistance, tripduration, startdate, starttime, enddate,	endtime, startlatitude, startlongitude, endlatitude, endlongitude, triproute, create_dt
FROM trips
WHERE tripdistance in (
    SELECT MIN(tripdistance)
    FROM trips)
LIMIT 1;

'''
result = engine.execute(query23)
lat = pd.read_sql(query23, con = engine)
lat.head()

sumdid  tripdistance  tripduration   startdate        starttime   
0  PoweredYZL2A   -20324803.8          44.0  2019-07-27  23:14:23.683333  \

      enddate          endtime  startlatitude  startlongitude  endlatitude   
0  2019-07-27  23:58:23.820000        36.1585        -86.7683       36.126  \

   endlongitude                                          triproute   
0      -86.7892  [(36.158517, -86.768224), (36.158624, -86.7682...  \

                create_dt  
0 2019-07-28 05:32:05.610

Powered2460 has weird mileage because the GPS pings at 0,0 lat and long.  PoweredYZK2A just shows negative distance.

In [ ]:
outliers = '''
SELECT *
FROM trips
WHERE (tripdistance <= 1 AND tripduration > 5) 
    OR (tripdistance >26400 AND tripduration > 120)
ORDER BY tripdistance
'''

out1 = pd.read_sql(outliers, con=engine)
out1

pubtimestamp companyname triprecordnum   
0     2019-07-27 23:56:41.940        Bird       BRD5324  \
1     2019-07-25 16:39:39.323        Bird        BRD485   
2     2019-07-26 02:20:01.733        Bird       BRD2398   
3     2019-07-29 21:33:20.407        Bird       BRD1539   
4     2019-06-09 20:30:28.920        Bird       BRD3786   
...                       ...         ...           ...   
33791 2019-06-14 10:53:43.947        Lyft        LFT646   
33792 2019-06-24 06:51:32.000        JUMP          JMP2   
33793 2019-05-20 22:02:08.773        SPIN        SPI304   
33794 2019-06-02 05:50:38.117        Lime       LIM1408   
33795 2019-06-02 15:13:29.800        SPIN        SPI314   

                                            sumdid  tripduration   
0                                     PoweredYZL2A      44.00000  \
1                                     PoweredI4464       9.00000   
2                                     PoweredVXSZ2      25.00000   
3                                     Powered4VCCV       6.00000   
4                                     PoweredNUASB      10.00000   
...                                            ...           ...   
33791                                Powered713485     754.42815   
33792  Powered8d26b649-c9af-505d-a2b1-7f289fa89ae6    1440.00000   
33793                               Powered4567425    1923.00000   
33794                         PoweredDIOKCEJNO2ZE2     120.40000   
33795                               Powered7147041     641.00000   

       tripdistance   startdate        starttime     enddate          endtime   
0     -2.032480e+07  2019-07-27  23:14:23.683333  2019-07-27  23:58:23.820000  \
1     -1.990092e+07  2019-07-25  16:31:14.886666  2019-07-25  16:40:29.046666   
2     -9.337271e+06  2019-07-26  01:55:59.876666  2019-07-26  02:20:39.063333   
3     -7.634515e+05  2019-07-29  21:29:25.016666  2019-07-29  21:34:57.306666   
4     -5.157480e+05  2019-06-09  20:21:09.820000  2019-06-09  20:31:09.060000   
...             ...         ...              ...         ...              ...   
33791  1.237631e+05  2019-06-13  22:19:18.136666  2019-06-14  10:53:43.826666   
33792  2.066592e+05  2019-06-21         16:01:41  2019-06-24         00:53:52   
33793  2.131529e+05  2019-05-19         13:55:41  2019-05-20         21:58:17   
33794  2.152238e+05  2019-06-02         03:00:37  2019-06-02         05:01:01   
33795  2.304167e+05  2019-06-02         04:29:22  2019-06-02         15:09:24   

       startlatitude  startlongitude  endlatitude  endlongitude   
0          36.158500      -86.768300    36.126000    -86.789200  \
1          36.146100      -86.782300    36.157200    -86.776600   
2          36.149200      -86.814200    36.149300    -86.814200   
3          36.156700      -86.765200    36.157000    -86.765200   
4          36.165700      -86.777900    36.161600    -86.778800   
...              ...             ...          ...           ...   
33791      36.160330      -86.778430    36.125030    -86.774690   
33792      36.152973      -86.783550    36.143887    -86.799610   
33793      36.153890      -86.794979    36.563954    -87.310677   
33794      36.160966      -86.778942    36.523999    -87.226165   
33795      36.160997      -86.778491    36.628313    -87.305048   

                                               triproute   
0      [(36.158517, -86.768224), (36.158624, -86.7682...  \
1      [(36.14634, -86.782481), (36.146169, -86.78243...   
2      [(36.149286, -86.814144), (36.149341, -86.8141...   
3                              [(36.157004, -86.765112)]   
4      [(36.165769, -86.777705), (36.165862, -86.7778...   
...                                                  ...   
33791  [(36.16033, -86.77843), (36.16006, -86.77847),...   
33792                                                 []   
33793  [(36.56395417641528, -87.31067726396705), (36....   
33794  [(36.16097, -86.77894), (36.16371, -86.78033),...   
33795  [(36.15839, -86.76854), (36.15839, -86.76854),

In [ ]:
query24 = '''
SELECT COUNT(sumdgroup)
FROM scooters
WHERE sumdgroup = 'bicycle';

'''
result = engine.execute(query24)
lat = pd.read_sql(query24, con = engine)
lat.head()

count
0  26529

How many companies are there and how many scooters does each company have?

In [ ]:
query25 = '''
SELECT companyname, COUNT(sumdid)
FROM scooters
WHERE sumdgroup <> 'bicycle' 
GROUP BY companyname;

'''
result = engine.execute(query25)
companies = pd.read_sql(query25, con = engine)
companies

companyname     count
0        Bird  12251590
1        Bolt   3477198
2      Gotcha   4679280
3        Jump  21835098
4        Lime  16497732
5        Lyft   9087043
6        Spin   5559573

In [ ]:
query26 = '''
SELECT companyname, COUNT(DISTINCT sumdid)
FROM scooters
WHERE sumdgroup <> 'bicycle' 
GROUP BY companyname;

'''
result = engine.execute(query26)
companies = pd.read_sql(query26, con = engine)
companies

companyname  count
0        Bird   3860
1        Bolt    360
2      Gotcha    224
3        Jump   1210
4        Lime   1818
5        Lyft   1735
6        Spin    805

companyname	count
0	Bird	3860
1	Bolt	360
2	Gotcha	224
3	Jump	1210
4	Lime	1818
5	Lyft	1735
6	Spin	805